bert/src/bert_layers.py

```diff
@@ -0,0 +1,859 @@
+# coding=utf-8
+# Copyright 2022 MosaicML Composer authors
+# Copyright 2018 The Google AI Language Team Authors and The HuggingFace Inc. team.
+# Copyright (c) 2018-2021, NVIDIA CORPORATION.  All rights reserved.
+# Copyright (c) 2022, Tri Dao.
+# SPDX-License-Identifier: Apache-2.0
+
+import copy
+import warnings
+import logging
+import math
+from typing import List, Optional, Tuple, Union
+
+import torch
+import torch.nn as nn
+from torch.nn.modules.utils import consume_prefix_in_state_dict_if_present
+from einops import rearrange
+from transformers.activations import ACT2FN
+from transformers.modeling_outputs import MaskedLMOutput, SequenceClassifierOutput
+from transformers.models.bert.modeling_bert import BertPredictionHeadTransform, BertPreTrainedModel
+
+from src.bert_padding import pad_input, unpad_input, unpad_input_only, index_first_axis, index_put_first_axis
+try:
+    from src.flash_attn_triton import flash_attn_qkvpacked_func
+except ImportError as e:
+    flash_attn_qkvpacked_func = None
+
+
+
+logger = logging.getLogger(__name__)
+
+
+class BertEmbeddings(nn.Module):
+    """Construct the embeddings from word, NOT position since we use ALiBi and token_type embeddings."""
+
+    def __init__(self, config):
+        super().__init__()
+        self.word_embeddings = nn.Embedding(config.vocab_size, config.hidden_size, padding_idx=config.pad_token_id)
+        # ALiBi doesn't use position embeddings
+        self.token_type_embeddings = nn.Embedding(config.type_vocab_size, config.hidden_size)
+
+        # self.LayerNorm is not snake-cased to stick with TensorFlow model variable name and be able to load
+        # any TensorFlow checkpoint file
+        self.LayerNorm = nn.LayerNorm(config.hidden_size, eps=config.layer_norm_eps)
+        self.dropout = nn.Dropout(config.hidden_dropout_prob)
+        self.register_buffer(
+            "token_type_ids", torch.zeros(config.max_position_embeddings, dtype=torch.long), persistent=False
+        )
+
+    def forward(
```

Comment:
Do we want more extensive documentation here?

Reply:
Good suggestion. Added more exposition throughout.